feat: refactor Plux architecture with modular design and improved feedback

navahas opened the pull request:
This PR introduces a major update to Plux, focusing on modularity, user feedback, and developer clarity.
The codebase is now structured in clear modules with improved error handling and a better UX through tmux messages.

---

### Changes

#### plux

* Added new modules: `config`, `error`, `git`
* Improved user feedback and logging during plugin installation
* Refactored main logic for cleaner architecture and better error propagation
* Enhanced tmux message display for runtime feedback
* Added `anyhow` dependency for unified error handling

#### murus

* Added `display_message` method

#### **docs**

* Updated `README.md` with installation instructions, configuration details, and TPM compatibility notes

---
### Motivation
* Improve maintainability and separation of concerns
* Provide better runtime feedback and error clarity
* Lay groundwork for future enhancements (plugin lifecycle management, logging improvements, etc.)

navahas committed:
fix(setup): create config and plugins directories on first run if not found

- Add clear installations steps in README.md to guide first-time users.

navahas committed:
chore: bump version to 0.2.0

navahas committed:
feat: automatically remove orphaned plugins on run

navahas committed:
Merge branch 'feat/plugin-remove' into trunk/master

navahas committed:
feat: add run feedback to user at tmux reload source

navahas committed:
feat: improve PLUX UX and modularize architecture

- murus:
    * add display_message_with_duration
- plux:
    * improve user feedback & logging
    * refactor modules/architecture (config, error, git)
    * orphan cleanup & concurrent sourcing
    * add `anyhow` dep
- docs:
    * update README (install from crates.io/source, configuration, execution modes, TPM migration)

navahas committed:
docs: update README.md for clearer installation instructions

nfejzic commented on plux/Cargo.toml:
```diff
@@ -8,8 +8,10 @@ license.workspace = true
 clap = { version = "4.5.48", features = ["derive"] }
-murus = { version = "0.1.1" }
+# murus = "0.1.1"
+murus = { path = "../murus" }
```
Could we pull out updates to `murus` into a separate PR so we can use the right version?

It wouldn't be a big deal if we don't, but it would help somewhat.

navahas replied:
Yeah, the reason for the local path was that I needed to add `display_message` and `display_message_with_duration` methods to murus. You will need to change the murus version once you publish/update the package.

nfejzic commented on plux/src/main.rs:
```diff
@@ -18,134 +16,172 @@ const HELP_TEMPLATE: &str = r#"
+ |     ___/    |   |    |   /\     / 
+ |    |   |    |___|    |  / /     \ 
+ |____|   |_______ \______/ /___/\  \
+                  \/              \_/
```
Hah, nice!

nfejzic commented on plux/src/config.rs:
```diff
@@ -0,0 +1,120 @@
+            source: e,
+        })?;
+
+        println!("Created default config file at {}", path.display());
+        println!("Add your plugins to this file and reload tmux configuration.");
```
Should we display this message using `display-message` command in `tmux`?

Or rather, what I'm asking is why not display this with `display-message`?

navahas replied:
So the reason I used `println!()` here is that `display-message` shows the message in the tmux status line (the single-line area at the bottom) and the message disappears after a timeout.

Since this is important setup info `println!()` makes more sense so it stays visible in the terminal output, also mimics the approach of TPM.

That said, this is tied to the first comment. Further, just to flex the plux logo is worth it.

nfejzic commented on plux/src/git.rs:
```diff
@@ -0,0 +1,126 @@
+            Ok(())
+        } else {
+            Err(GitError::CommandFailed {
+                command: "fetch --all --tags".to_string(),
```
Nitpick: I recognize a pattern for error reporting that we could simplify a little bit. Basically, the command arguments are repeated when reporting the error. We could take advantage of [slice::join](https://doc.rust-lang.org/std/primitive.slice.html#method.join) in these functions: 


```suggestion
        let args = ["fetch", "--all", "--tags"];
        let output = self
            .command()
            .args(&args)
            .output()
            .map_err(GitError::IoError)?;

        if output.status.success() {
            Ok(())
        } else {
            Err(GitError::CommandFailed {
                command: args.join(" "),
```

navahas replied:
Good call! we should also apply this pattern to the others: `fetch_tags`, `checkout`, and `get_default_branch`. 

Then, `clone_shallow` can be left as-is since it has the dynamic url arg appended separately, and following the pattern would demand the quirk `.to_string_lossy()` to get the &str

nfejzic commented on plux/src/main.rs:
```diff
@@ -18,134 +16,172 @@ const HELP_TEMPLATE: &str = r#"
+                eprintln!("\nTroubleshooting:");
+                eprintln!("  1. Check TOML syntax is valid");
+                eprintln!("  2. Ensure [plugins] section exists");
+                eprintln!("  3. See example format in {}", path.display());
```
Can we guarantee that there is example in file at `path`?

navahas replied:
We can guarantee that plux will create a working config file, but the error if the parsing goes wrong is not properly handled. Need to change that

nfejzic commented on plux/src/git.rs:
```diff
@@ -0,0 +1,126 @@
+    /// Create a new Git instance for running commands
+    pub fn new() -> Self {
+        Self { repo_path: None }
+    }
```
I would drop the redundant `new()` function and use `impl Default` wherever needed.

navahas replied:
Makes sense, this is redundant.

nfejzic reviewed (commented):
Hey @navahas, thank you so much for your contributions!

My notification settings for repositories were not set to notify me about everything, so I missed this. Sorry for getting back to you months later 🫣. Please, feel free to tag me if you decide to contribute again! 

I've reviewed this PR, and I've seen some others but those are superseded by this one right? If so, we can close those and get this one merged. 

The PR is great! I've left some nitpicky comments, please take a look. If you don't have time or don't feel eager to work on them please let me know. We can merge this and I can handle them myself later as well 😄. Thanks again!

navahas committed:
chore: remove invisible char from logo

navahas commented:
I'll make a new draft to cover the previous comments this week (most likely tomorrow). No worries for the delay!

navahas committed:
feat: address PR#20 review, git error reporting pattern, improve error output

navahas commented:
Hopefully the previous commit addresses all the comments, just let me know if any further thing is needed.

A preview of the updated error output for bad git remotes or toml config.

<img width="380" alt="image" src="https://github.com/user-attachments/assets/c71b492a-d4ce-4378-9033-ac343aa72c0c" />
<img width="380" alt="image" src="https://github.com/user-attachments/assets/3d5ebe9f-7583-47cb-abb7-c6b4c610eda3" />

Edit: Forgot to mention that `plux` version should also be updated!

nfejzic committed:
chore: format files

nfejzic reviewed (approved):
Thanks @navahas

nfejzic committed:
chore: resolve clippy warnings

nfejzic committed:
chore: remove `anyhow` dependency
